File: lumiere-backend/app/utils/budget_estimator.py
from app.utils.measurements import (
    EMPTY_BUDGET_TOTALS,
    get_ceiling_area,
    get_floor_area,
    get_room_id_for_position,
    get_wall_area,
)


CATEGORY_KEYS = {
    "furniture": "furniture",
    "decor": "decor",
    "wall": "walls",
    "floor": "floor",
    "ceiling": "ceiling",
    "door": "doors",
    "window": "windows",
    "light": "lights",
}

TOTAL_KEYS = {
    "furniture": "furnitureTotal",
    "decor": "decorTotal",
    "wall": "wallTotal",
    "floor": "floorTotal",
    "ceiling": "ceilingTotal",
    "door": "doorTotal",
    "window": "windowTotal",
    "light": "lightTotal",
}

AREA_COST_TARGET_TYPES = {"wall", "floor", "ceiling"}
DEFAULT_COST_SOURCE = "manual"


def _finite_number(value, fallback=0):
    return value if isinstance(value, (int, float)) and not isinstance(value, bool) else fallback


def _round_money(value):
    return round(value, 2) if isinstance(value, (int, float)) and not isinstance(value, bool) else 0


def _is_decor_item(item):
    category = str(item.get("category") or item.get("type") or "").lower()
    return "decor" in category or "rug" in category or "mirror" in category


def _valid_rules(rules):
    return [
        rule for rule in (rules if isinstance(rules, list) else [])
        if isinstance(rule, dict) and rule.get("targetType") and rule.get("scope")
    ]


def _build_budget_quotation(scene_data, summary, project_name=None, generated_at=None):
    budget = scene_data.get("budget") if isinstance(scene_data.get("budget"), dict) else {}
    existing_quotation = budget.get("quotation") if isinstance(budget.get("quotation"), dict) else {}
    return {
        "projectName": project_name or scene_data.get("projectName") or scene_data.get("title") or scene_data.get("name") or "Untitled Room",
        "date": generated_at or existing_quotation.get("date"),
        "currency": budget.get("currency") or "AED",
        "roomBreakdown": summary["byRoom"],
        "lineItems": [
            {
                "id": item["targetId"],
                "targetId": item["targetId"],
                "targetType": item["targetType"],
                "label": item.get("label"),
                "roomId": item.get("roomId"),
                "quantity": item.get("quantity"),
                "areaSqm": item.get("areaSqm"),
                "unitPrice": item.get("amount"),
                "unit": item.get("unit"),
                "pricingMode": item.get("pricingMode"),
                "costSource": item.get("costSource"),
                "appliedRuleId": item.get("appliedRuleId"),
                "total": item.get("calculatedCost"),
                "customOverride": item.get("customOverride"),
            }
            for item in summary["breakdown"]
        ],
        "totals": summary["totals"],
        "grandTotal": summary["grandTotal"],
    }


def _find_applied_rule(element, rules):
    target_type = element["targetType"]

    if target_type in AREA_COST_TARGET_TYPES:
        for scope in ("singleItem", "roomType", "globalType"):
            for rule in rules:
                if rule.get("targetType") != target_type or rule.get("scope") != scope:
                    continue
                if scope == "singleItem" and rule.get("targetId") == element.get("targetId"):
                    return rule
                if scope == "roomType" and rule.get("roomId") == element.get("roomId"):
                    return rule
                if scope == "globalType":
                    return rule
        return None

    for scope in ("singleItem", "globalType"):
        for rule in rules:
            if rule.get("targetType") != target_type or rule.get("scope") != scope:
                continue
            if scope == "singleItem" and rule.get("targetId") == element.get("targetId"):
                return rule
            if scope == "globalType":
                return rule
    return None


def _collect_budget_elements(scene_data):
    rooms = scene_data.get("rooms") if isinstance(scene_data.get("rooms"), list) else []
    walls = scene_data.get("walls") if isinstance(scene_data.get("walls"), list) else []
    furniture = scene_data.get("furniture")
    if furniture is None:
        furniture = scene_data.get("placedItems")
    furniture = furniture if isinstance(furniture, list) else []
    lighting = scene_data.get("lighting") if isinstance(scene_data.get("lighting"), dict) else {}
    placed_lights = lighting.get("placedLights") if isinstance(lighting.get("placedLights"), list) else []
    floor = scene_data.get("materials", {}).get("floor") or scene_data.get("floorMaterial") or {}
    ceiling = scene_data.get("materials", {}).get("ceiling") or scene_data.get("ceilingMaterial") or {}

    elements = []

    for room in rooms:
        if not isinstance(room, dict) or not room.get("id"):
            continue
        measurements = room.get("measurements") if isinstance(room.get("measurements"), dict) else {}
        room_label = room.get("name") or room.get("label") or "Room"
        elements.append({
            "targetId": f"floor:{room['id']}",
            "targetType": "floor",
            "roomId": room["id"],
            "label": f"{room_label} floor",
            "areaSqm": _finite_number(measurements.get("floorAreaSqm"), get_floor_area(floor, room)),
        })
        elements.append({
            "targetId": f"ceiling:{room['id']}",
            "targetType": "ceiling",
            "roomId": room["id"],
            "label": f"{room_label} ceiling",
            "areaSqm": _finite_number(measurements.get("ceilingAreaSqm"), get_ceiling_area(ceiling, room)),
        })

    for wall in walls:
        if not isinstance(wall, dict) or not wall.get("id"):
            continue
        measurements = wall.get("measurements") if isinstance(wall.get("measurements"), dict) else {}
        elements.append({
            "targetId": wall["id"],
            "targetType": "wall",
            "roomId": wall.get("roomId"),
            "label": wall.get("label") or "Wall",
            "areaSqm": _finite_number(measurements.get("areaSqm"), get_wall_area(wall)),
        })

        for door in wall.get("doors") if isinstance(wall.get("doors"), list) else []:
            if not isinstance(door, dict) or not door.get("id"):
                continue
            measurements = door.get("measurements") if isinstance(door.get("measurements"), dict) else {}
            elements.append({
                "targetId": door["id"],
                "targetType": "door",
                "roomId": door.get("roomId") or wall.get("roomId"),
                "label": door.get("label") or "Door",
                "quantity": _finite_number(measurements.get("quantity"), 1),
            })

        for window in wall.get("windows") if isinstance(wall.get("windows"), list) else []:
            if not isinstance(window, dict) or not window.get("id"):
                continue
            measurements = window.get("measurements") if isinstance(window.get("measurements"), dict) else {}
            elements.append({
                "targetId": window["id"],
                "targetType": "window",
                "roomId": window.get("roomId") or wall.get("roomId"),
                "label": window.get("label") or "Window",
                "quantity": _finite_number(measurements.get("quantity"), 1),
            })

    for item in furniture:
        if not isinstance(item, dict) or not item.get("id"):
            continue
        measurements = item.get("measurements") if isinstance(item.get("measurements"), dict) else {}
        target_type = "decor" if _is_decor_item(item) else "furniture"
        elements.append({
            "targetId": item["id"],
            "targetType": target_type,
            "roomId": item.get("roomId") or get_room_id_for_position(rooms, item.get("position") or [0, 0, 0]),
            "label": item.get("label") or item.get("name") or item.get("filename") or "Furniture",
            "quantity": _finite_number(measurements.get("quantity"), 1),
        })

    for light in placed_lights:
        if not isinstance(light, dict) or not light.get("id"):
            continue
        measurements = light.get("measurements") if isinstance(light.get("measurements"), dict) else {}
        elements.append({
            "targetId": light["id"],
            "targetType": "light",
            "roomId": light.get("roomId") or get_room_id_for_position(rooms, light.get("position") or [0, 0, 0]),
            "label": light.get("label") or f"{light.get('budgetCategory') or light.get('type') or 'Light'} light",
            "lightType": light.get("type"),
            "budgetCategory": light.get("budgetCategory"),
            "quantity": _finite_number(measurements.get("quantity"), _finite_number(light.get("quantity"), 1)),
        })

    return elements


def calculate_budget_summary(scene_data, budget_rules, project_name=None, generated_at=None):
    scene_data = scene_data if isinstance(scene_data, dict) else {}
    rules = _valid_rules(budget_rules)
    by_category = {
        "furniture": 0,
        "decor": 0,
        "walls": 0,
        "floor": 0,
        "ceiling": 0,
        "doors": 0,
        "windows": 0,
        "lights": 0,
    }
    by_room = {}
    totals = {**EMPTY_BUDGET_TOTALS}
    breakdown = []

    for element in _collect_budget_elements(scene_data):
        applied_rule = _find_applied_rule(element, rules)
        is_area_based = element["targetType"] in AREA_COST_TARGET_TYPES
        quantity = _finite_number(element.get("areaSqm") if is_area_based else element.get("quantity"), 1)
        amount = _finite_number(applied_rule.get("amount") if applied_rule else None)
        calculated_cost = _round_money(quantity * amount)
        category_key = CATEGORY_KEYS.get(element["targetType"])
        total_key = TOTAL_KEYS.get(element["targetType"])

        if category_key:
            by_category[category_key] = _round_money(by_category.get(category_key, 0) + calculated_cost)
        if total_key:
            totals[total_key] = _round_money(totals.get(total_key, 0) + calculated_cost)
        if element.get("roomId"):
            by_room[element["roomId"]] = _round_money(by_room.get(element["roomId"], 0) + calculated_cost)

        breakdown.append({
            **element,
            "quantity": None if is_area_based else quantity,
            "areaSqm": quantity if is_area_based else None,
            "appliedRuleId": applied_rule.get("id") if applied_rule else None,
            "pricingMode": applied_rule.get("pricingMode") if applied_rule else None,
            "unit": applied_rule.get("unit") if applied_rule else None,
            "costSource": (applied_rule.get("costSource") or DEFAULT_COST_SOURCE) if applied_rule else None,
            "amount": amount,
            "calculatedCost": calculated_cost,
            "customOverride": applied_rule.get("scope") == "singleItem" if applied_rule else False,
        })

    grand_total = _round_money(sum(by_category.values()))
    totals["grandTotal"] = grand_total

    summary = {
        "grandTotal": grand_total,
        "byCategory": by_category,
        "byRoom": by_room,
        "breakdown": breakdown,
        "totals": totals,
        "snapshots": [
            {
                "targetId": item["targetId"],
                "targetType": item["targetType"],
                "roomId": item.get("roomId"),
                "quantity": item.get("quantity"),
                "areaSqm": item.get("areaSqm"),
                "appliedRuleId": item.get("appliedRuleId"),
                "calculatedCost": item["calculatedCost"],
                "costSource": item.get("costSource"),
                "customOverride": item["customOverride"],
            }
            for item in breakdown
        ],
    }
    return {
        **summary,
        "quotation": _build_budget_quotation(scene_data, summary, project_name, generated_at),
    }

File: lumiere-frontend/src/utils/budgetEstimator.js
import { AREA_COST_TARGET_TYPES, BUDGET_COST_SOURCES, EMPTY_BUDGET_TOTALS } from './budgetContract';
import {
  getCeilingArea,
  getFloorArea,
  getRoomIdForPosition,
  getWallArea,
} from './measurements';

const CATEGORY_KEYS = Object.freeze({
  furniture: 'furniture',
  decor: 'decor',
  wall: 'walls',
  floor: 'floor',
  ceiling: 'ceiling',
  door: 'doors',
  window: 'windows',
  light: 'lights',
});

const TOTAL_KEYS = Object.freeze({
  furniture: 'furnitureTotal',
  decor: 'decorTotal',
  wall: 'wallTotal',
  floor: 'floorTotal',
  ceiling: 'ceilingTotal',
  door: 'doorTotal',
  window: 'windowTotal',
  light: 'lightTotal',
});

function finiteNumber(value, fallback = 0) {
  return Number.isFinite(value) ? value : fallback;
}

function roundMoney(value) {
  return Number.isFinite(value) ? Number(value.toFixed(2)) : 0;
}

function isDecorItem(item = {}) {
  const category = String(item.category ?? item.type ?? '').toLowerCase();
  return category.includes('decor') || category.includes('rug') || category.includes('mirror');
}

function getRules(rules = []) {
  return Array.isArray(rules) ? rules.filter((rule) => rule?.targetType && rule?.scope) : [];
}

function buildBudgetQuotation(scene, summary) {
  const projectName = scene.projectName ?? scene.title ?? scene.name ?? 'Untitled Room';
  const currency = scene.budget?.currency ?? 'AED';
  return {
    projectName,
    date: scene.budget?.quotation?.date ?? new Date().toISOString(),
    currency,
    roomBreakdown: summary.byRoom,
    lineItems: summary.breakdown.map((item) => ({
      id: item.targetId,
      targetId: item.targetId,
      targetType: item.targetType,
      label: item.label,
      roomId: item.roomId ?? null,
      quantity: item.quantity,
      areaSqm: item.areaSqm,
      unitPrice: item.amount,
      unit: item.unit,
      pricingMode: item.pricingMode,
      costSource: item.costSource,
      appliedRuleId: item.appliedRuleId,
      total: item.calculatedCost,
      customOverride: item.customOverride,
    })),
    totals: summary.totals,
    grandTotal: summary.grandTotal,
  };
}

function findAppliedRule(element, rules) {
  if (AREA_COST_TARGET_TYPES.includes(element.targetType)) {
    return (
      rules.find((rule) => (
        rule.scope === 'singleItem' &&
        rule.targetType === element.targetType &&
        rule.targetId === element.targetId
      )) ??
      rules.find((rule) => (
        rule.scope === 'roomType' &&
        rule.targetType === element.targetType &&
        rule.roomId === element.roomId
      )) ??
      rules.find((rule) => (
        rule.scope === 'globalType' &&
        rule.targetType === element.targetType
      )) ??
      null
    );
  }

  return (
    rules.find((rule) => (
      rule.scope === 'singleItem' &&
      rule.targetType === element.targetType &&
      rule.targetId === element.targetId
    )) ??
    rules.find((rule) => (
      rule.scope === 'globalType' &&
      rule.targetType === element.targetType
    )) ??
    null
  );
}

function getSceneRooms(scene = {}) {
  return Array.isArray(scene.rooms) ? scene.rooms : [];
}

function getSceneFurniture(scene = {}) {
  if (Array.isArray(scene.furniture)) return scene.furniture;
  return Array.isArray(scene.placedItems) ? scene.placedItems : [];
}

function collectBudgetElements(scene = {}) {
  const rooms = getSceneRooms(scene);
  const walls = Array.isArray(scene.walls) ? scene.walls : [];
  const furniture = getSceneFurniture(scene);
  const placedLights = Array.isArray(scene.lighting?.placedLights) ? scene.lighting.placedLights : [];
  const floorMaterial = scene.materials?.floor ?? scene.floorMaterial ?? {};
  const ceilingMaterial = scene.materials?.ceiling ?? scene.ceilingMaterial ?? {};

  const elements = [];

  rooms.forEach((room) => {
    if (!room?.id) return;
    const roomMeasurements = room.measurements ?? {};
    elements.push({
      targetId: `floor:${room.id}`,
      targetType: 'floor',
      roomId: room.id,
      label: `${room.name ?? room.label ?? 'Room'} floor`,
      areaSqm: finiteNumber(roomMeasurements.floorAreaSqm, getFloorArea(floorMaterial, room)),
    });
    elements.push({
      targetId: `ceiling:${room.id}`,
      targetType: 'ceiling',
      roomId: room.id,
      label: `${room.name ?? room.label ?? 'Room'} ceiling`,
      areaSqm: finiteNumber(roomMeasurements.ceilingAreaSqm, getCeilingArea(ceilingMaterial, room)),
    });
  });

  walls.forEach((wall) => {
    if (!wall?.id) return;
    elements.push({
      targetId: wall.id,
      targetType: 'wall',
      roomId: wall.roomId ?? null,
      label: wall.label ?? 'Wall',
      areaSqm: finiteNumber(wall.measurements?.areaSqm, getWallArea(wall)),
    });

    (wall.doors ?? []).forEach((door) => {
      if (!door?.id) return;
      elements.push({
        targetId: door.id,
        targetType: 'door',
        roomId: door.roomId ?? wall.roomId ?? null,
        label: door.label ?? 'Door',
        quantity: finiteNumber(door.measurements?.quantity, 1),
      });
    });

    (wall.windows ?? []).forEach((window) => {
      if (!window?.id) return;
      elements.push({
        targetId: window.id,
        targetType: 'window',
        roomId: window.roomId ?? wall.roomId ?? null,
        label: window.label ?? 'Window',
        quantity: finiteNumber(window.measurements?.quantity, 1),
      });
    });
  });

  furniture.forEach((item) => {
    if (!item?.id) return;
    const targetType = isDecorItem(item) ? 'decor' : 'furniture';
    elements.push({
      targetId: item.id,
      targetType,
      roomId: item.roomId ?? getRoomIdForPosition(rooms, item.position),
      label: item.label ?? item.name ?? item.filename ?? 'Furniture',
      quantity: finiteNumber(item.measurements?.quantity, 1),
    });
  });

  placedLights.forEach((light) => {
    if (!light?.id) return;
    elements.push({
      targetId: light.id,
      targetType: 'light',
      roomId: light.roomId ?? getRoomIdForPosition(rooms, light.position),
      label: light.label ?? `${light.budgetCategory ?? light.type ?? 'Light'} light`,
      lightType: light.type ?? null,
      budgetCategory: light.budgetCategory ?? null,
      quantity: finiteNumber(light.measurements?.quantity, finiteNumber(light.quantity, 1)),
    });
  });

  return elements;
}

export function calculateBudgetSummary(scene = {}, rules = []) {
  const validRules = getRules(rules);
  const byCategory = {
    furniture: 0,
    decor: 0,
    walls: 0,
    floor: 0,
    ceiling: 0,
    doors: 0,
    windows: 0,
    lights: 0,
  };
  const byRoom = {};
  const totals = { ...EMPTY_BUDGET_TOTALS };

  const breakdown = collectBudgetElements(scene).map((element) => {
    const appliedRule = findAppliedRule(element, validRules);
    const isAreaBased = AREA_COST_TARGET_TYPES.includes(element.targetType);
    const quantity = isAreaBased ? finiteNumber(element.areaSqm) : finiteNumber(element.quantity, 1);
    const amount = finiteNumber(appliedRule?.amount);
    const calculatedCost = roundMoney(quantity * amount);
    const categoryKey = CATEGORY_KEYS[element.targetType];
    const totalKey = TOTAL_KEYS[element.targetType];

    if (categoryKey) byCategory[categoryKey] = roundMoney((byCategory[categoryKey] ?? 0) + calculatedCost);
    if (totalKey) totals[totalKey] = roundMoney((totals[totalKey] ?? 0) + calculatedCost);
    if (element.roomId) byRoom[element.roomId] = roundMoney((byRoom[element.roomId] ?? 0) + calculatedCost);

    return {
      ...element,
      quantity: isAreaBased ? undefined : quantity,
      areaSqm: isAreaBased ? quantity : undefined,
      appliedRuleId: appliedRule?.id ?? null,
      pricingMode: appliedRule?.pricingMode ?? null,
      unit: appliedRule?.unit ?? null,
      costSource: appliedRule?.costSource ?? (appliedRule ? BUDGET_COST_SOURCES.MANUAL : null),
      amount,
      calculatedCost,
      customOverride: appliedRule?.scope === 'singleItem',
    };
  });

  const grandTotal = roundMoney(Object.values(byCategory).reduce((sum, value) => sum + value, 0));
  totals.grandTotal = grandTotal;

  const summary = {
    grandTotal,
    byCategory,
    byRoom,
    breakdown,
    totals,
    snapshots: breakdown.map((item) => ({
      targetId: item.targetId,
      targetType: item.targetType,
      roomId: item.roomId,
      quantity: item.quantity,
      areaSqm: item.areaSqm,
      appliedRuleId: item.appliedRuleId,
      calculatedCost: item.calculatedCost,
      costSource: item.costSource,
      customOverride: item.customOverride,
    })),
  };
  return {
    ...summary,
    quotation: buildBudgetQuotation(scene, summary),
  };
}
